fdbserver/DataDistribution.actor.cpp

```diff
@@ -1512,219 +1631,690 @@ ACTOR Future<Void> ddGetMetrics(GetDataDistributorMetricsRequest req,
 	while (begin < range.end) {
 		currentRange = KeyRangeRef(begin, range.end);
-		std::vector<AuditStorageState> auditStates =
-		    wait(getAuditStateForRange(self->txnProcessor->context(), audit->id, currentRange));
+		if (makeProgressbyServer) {
+			ASSERT(audit->coreState.getType() == AuditType::ValidateStorageServerShard);
+			wait(store(auditStates,
+			           getAuditStateByServer(self->txnProcessor->context(),
+			                                 audit->coreState.getType(),
+			                                 audit->coreState.id,
+			                                 serverId,
+			                                 currentRange)));
+		} else {
+			ASSERT(audit->coreState.getType() == AuditType::ValidateLocationMetadata);
+			wait(store(
+			    auditStates,
+			    getAuditStateByRange(
+			        self->txnProcessor->context(), audit->coreState.getType(), audit->coreState.id, currentRange)));
+		}
 		ASSERT(!auditStates.empty());
 		begin = auditStates.back().range.end;
+		TraceEvent(SevInfo, "DDAuditMakeProgressOnServerDispatch", self->ddId)
+		    .detail("ServerID", serverId)
+		    .detail("AuditID", audit->coreState.id)
+		    .detail("CurrentRange", currentRange)
+		    .detail("AuditType", audit->coreState.getType())
+		    .detail("NextBegin", begin)
+		    .detail("RangeEnd", range.end);
 		for (const auto& auditState : auditStates) {
 			const AuditPhase phase = auditState.getPhase();
-			audit->auditMap.insert(auditState.range, phase);
+			ASSERT(phase != AuditPhase::Running && phase != AuditPhase::Failed);
+			totalCount++;
 			if (phase == AuditPhase::Complete) {
-				continue;
+				completedCount++;
 			} else if (phase == AuditPhase::Error) {
-				throw audit_storage_error();
+				completedCount++;
+				audit->foundError = true;
 			} else {
-				audit->actors.add(scheduleAuditForRange(self, audit, auditState.range));
+				ASSERT(phase == AuditPhase::Invalid);
+				AuditStorageRequest req(audit->coreState.id, auditState.range, audit->coreState.getType());
+				audit->actors.add(doAuditOnStorageServer(self, audit, ssi, req));
 			}
 		}
 		wait(delay(0.1));
 	}
-
+	TraceEvent(SevInfo, "DDAuditMakeProgressOnServerEnd", self->ddId)
+	    .detail("ServerID", serverId)
+	    .detail("AuditID", audit->coreState.id)
+	    .detail("Range", range)
+	    .detail("AuditType", audit->coreState.getType())
+	    .detail("TotalRanges", totalCount)
+	    .detail("TotalComplete", completedCount)
+	    .detail("CompleteRatio", completedCount * 1.0 / totalCount);
 	return Void();
 }
 
-ACTOR Future<Void> scheduleAuditForRange(Reference<DataDistributor> self,
-                                         std::shared_ptr<DDAudit> audit,
-                                         KeyRange range) {
-	TraceEvent(SevInfo, "DDScheduleAuditForRangeBegin", self->ddId)
-	    .detail("AuditID", audit->id)
+// Schedule audit task on the input range
+ACTOR Future<Void> auditMakeProgressOnRange(Reference<DataDistributor> self,
```

Comment:
nit: the function name usually starts with a verb

Reply:
Understand. Thanks!